tools/ansible/tasks/wsu/main.yaml

```diff
@@ -2,29 +2,131 @@
+      shell: |
+        echo "https://github.com/kubernetes/kubernetes/blob/release-"\
+        "{{ kubernetes_version_without_patch.stdout }}/CHANGELOG/CHANGELOG-"\
+        "{{ kubernetes_version_without_patch.stdout }}.md"
+      register: changelog
+
+    - name: Get kube node binary SHA
+      shell: |
+        curl -s "{{ changelog.stdout }}" | \
+        grep -A 1 "{{ kubernetes_version.stdout }}" | \
+        grep -A 1 node-windows | tail -n 1 | sed -e 's/.*<code>\(.*\)<\/code>.*/\1/'
+      register: kube_node_sha
+
+    # OpenShift cluster version
+    - name: Register cluster version
+      shell: "echo {{ map_kubernetes_to_openshift_version[kubernetes_version_without_patch.stdout] }}"
+      register: cluster_version
 
-        - name: Move WMCB to temporary directory
-          command: mv "{{ project_root }}/{{ wmcb_exe }}" "{{ tmp_dir.path }}/{{ wmcb_exe }}"
+    # curl -s https://api.github.com/repos/openshift/windows-machine-config-operator/releases gets the list of all
+    # releases, we use the cluster version from previous ansible task to select the release we need.
+    - name: Get release
+      shell: |
+        RELEASES="$(curl -s https://api.github.com/repos/openshift/windows-machine-config-operator/releases)"
+        TAGNAME="$(echo $RELEASES | jq '.[]' | jq '.tag_name' | \
+        grep -i {{ cluster_version.stdout }} |\
+        sort -r | awk 'NR==1')"
+        echo $RELEASES | \
+        jq '.[] | select (.tag_name == '$TAGNAME')'
+      register: release
+
+    - name: Gather required files
+      block:
+        - name: Build WMCB
+          when: build_wmcb is defined and build_wmcb
+          block:
+            - name: Build WMCB
+              make:
+                target: build
+                chdir: "{{ project_root }}"
+
+            - name: Move WMCB to temporary directory
+              command: mv "{{ project_root }}/{{ wmcb_exe }}" "{{ tmp_dir.path }}/{{ wmcb_exe }}"
+
+        - name: Download WMCB binary
+          when: not build_wmcb or build_wmcb is not defined
+          block:
+            # The wmcb release tags are supposed to conform to the following semver format:
+            # v<openshift-major-version>.<openshift-minor-version>.<wmcb_build_number>-<release_type>. For example, v4.4-1-alpha.
+            # For a <openshift-major-version>.<openshift-minor-version>, there can be many wmcb builds available, for
+            # example, OpenShift 4.4 can map to v4.4.1-alpha and v4.4.2-alpha, we will always pick up the latest build available
+            # which is v4.4.2-alpha,
+            # In future, when we cut beta and GA releases, the playbook should always go for GA version when available.
+            - name: Get wmcb url
+              shell: |
+                echo '{{ release.stdout }}' | \
+                jq ' .assets[] | select (.name == "{{ wmcb_exe }}" )' | \
+                jq -r .browser_download_url
+              register: wmcb_download_url
```

Comment:
There are multiple places where you are looking at `release.stdout`. What happens if the release is malformed and you don't the expected value? Do you have guards against this down the line?

Reply:
I have added a guard to ensure we have correct cluster version (by checking the presence of kubernetes version in the map) and I am checking if the release object was defined and not empty before proceeding

Reply:
If the `release.stdout` is malformed, it could potentially have data but just not in the format you expect. In those scenarios the above code will fail. That will result in variables like `wmcb_download_url` having incorrect or no data.

Reply:
Since I only moved this task from `win` host to `localhost` without making any significant changes, I think this is beyond the scope of this ticket

Reply:
That is a fair statement. It just means that you need to ensure that the guard you added for `release.stdout` is good enough which I don't think is the case.

Reply:
The playbook will fail when the actual download happens here https://github.com/openshift/windows-machine-config-bootstrapper/pull/148/files#diff-def35121485b59e044decc98a81c6402R118 and increasing the ansible log-level will help in getting the error. We can have some checks at get_url or let run ansible in dry-run mode to get the info.